src/plonk/circuit.rs

```diff
@@ -1072,11 +1156,280 @@ impl<F: Field> ConstraintSystem<F> {
         });
     }
 
-    /// Allocate a new selector.
+    /// This will compress selectors together depending on their provided
+    /// assignments. This `ConstraintSystem` will then be modified to add new
+    /// fixed columns (representing the actual selectors) and will return the
+    /// polynomials for those columns. Finally, an internal map is updated to
+    /// find which fixed column corresponds with a given `Selector`.
+    ///
+    /// Do not call this twice. Yes, this should be a builder pattern instead.
+    pub(crate) fn compress_selectors(mut self, selectors: Vec<Vec<bool>>) -> (Self, Vec<Vec<F>>) {
+        if selectors.is_empty() {
+            return (self, vec![]);
+        }
+        // The length of all provided selectors must be the same.
+        let n = selectors[0].len();
+        assert!(selectors.iter().all(|a| a.len() == n));
+        // and the number of provided selectors must be the same.
+        assert_eq!(selectors.len(), self.num_selectors);
+
+        // Compute the maximal degree of every simple selector. We only consider
+        // the expressions in gates, as lookup arguments cannot support simple
+        // selectors.
+        let mut degrees = vec![0; selectors.len()];
+        for expr in self.gates.iter().flat_map(|gate| gate.polys.iter()) {
+            let degree = expr.degree();
+            if let Some(selector) = expr.extract_simple_selector() {
+                degrees[selector.0] = max(degrees[selector.0], degree);
+            }
+        }
```

Comment:
At this point we should stable-sort the selectors by decreasing degree.
> Rationale: The number of selectors in a combination is limited by the maximum degree of any polynomial constraint for those selectors. The rest of the algorithm will attempt to combine selectors by iterating over them in order. If this order is such that we prefer to combine selectors for which the maximum degree is about the same, intuitively we should be able to combine more of them. In practice this leads to slightly better results, i.e. fewer combinations are typically needed.

Reply:
I was unable to find an example where sorting helped.